core/src/test/java/org/elasticsearch/gateway/PrimaryShardAllocatorTests.java

```diff
@@ -251,12 +251,31 @@ public void testAllocateToTheHighestVersionOnLegacyIndex() {
     }
 
     /**
+     * Tests that shard with allocation id is chosen if available in version-based allocation.
+     */
+    public void testVersionBasedAllocationPrefersShardWithAllocationId() {
+        RoutingAllocation allocation = routingAllocationWithOnePrimaryNoReplicas(yesAllocationDeciders(), false, Version.V_2_0_0);
+        testAllocator.addData(node1, 10, null, randomBoolean());
+        testAllocator.addData(node2, ShardStateMetaData.NO_VERSION, "some allocId", randomBoolean());
+        testAllocator.addData(node3, 12, null, randomBoolean());
+        boolean changed = testAllocator.allocateUnassigned(allocation);
+        assertThat(changed, equalTo(true));
+        assertThat(allocation.routingNodes().unassigned().ignored().isEmpty(), equalTo(true));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).size(), equalTo(1));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).currentNodeId(), equalTo(node2.id()));
+        assertThat(allocation.routingNodes().shardsWithState(ShardRoutingState.INITIALIZING).get(0).allocationId().getId(), equalTo("some allocId"));
+    }
+
+    /**
      * Tests that when restoring from a snapshot and we find a node with a shard copy and allocation
      * deciders say yes, we allocate to that node.
      */
     public void testRestore() {
-        RoutingAllocation allocation = getRestoreRoutingAllocation(yesAllocationDeciders());
-        testAllocator.addData(node1, 1, randomFrom(null, "allocId"), randomBoolean());
+        boolean hasActiveAllocation = randomBoolean();
+        String allocationId = hasActiveAllocation ? "allocId" : null;
+        long legacyVersion = hasActiveAllocation ? ShardStateMetaData.NO_VERSION : 1;
+        RoutingAllocation allocation = getRestoreRoutingAllocation(yesAllocationDeciders(), hasActiveAllocation);
```

Comment:
> This test is not only for cluster crash upon restore.

understood! the question is whether we see a valid scenario where we restore to an index where there are no active allocation ids but the shard state reports an allocation and no version. I think that's possible and thus we should test it?